## Lesson Goals
- Examine data for potential issues.
- Identify and fill in missing values.
- Identify and correct incorrect values.
- Remove low variance columns.
- Identify potential outliers.
- Correct incorrect data types.
- Remove special characters and clean categorical variables.
- Identify and remove duplicate records.


When working with data sets, you will find that they often require a bit of cleaning. Whether Pandas originally read the data types incorrectly, records are duplicated, the data contains special characters or missing value, or there are slightly different references to the same entity, every data analyst must know how to clean the data they are working with before analyzing it. In this lesson, you will learn about some of the most common problems that make data messy and methods for correcting those problems and cleaning your data.

The data set we are going to be using for this lesson is a messy version of the vehicles data set we worked with in the previous lesson. Let's import this version of our data set so that we can then practice cleaning it up.

In [1]:
import pandas as pd
import numpy as np
data = pd.read_csv("/Users/jon/Documents/IRONHACK/IronHackNotes/Data Cleaning/vehicles_messy.csv")

/var/folders/1k/45vk_vqn0z174f0vwyd90vqr0000gn/T/ipykernel_1360/2172240756.py:3: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/Users/jon/Documents/IRONHACK/IronHackNotes/Data Cleaning/vehicles_messy.csv")


## Examining Data for Potential Issues

One of the first things we want to do is examine the data and look for any potential issues. Some of the things we are interested in identifying in the data at this stage include:

- Missing values
- Special characters
- Incorrect values
- Extreme values or outliers
- Duplicate records
- Incorrect data types

The presence of these may cause problems when it's time to analyze the data, so we want to make sure we address them beforehand. We can start by visually inspecting the data using the head method, which will show us the first 5 rows of data.

In [6]:
data.head(5)


barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0  15.695714         0.0        0.0        0.0      19      0.0        0   
1  29.964545         0.0        0.0        0.0       9      0.0        0   
2  12.207778         0.0        0.0        0.0      23      0.0        0   
3  29.964545         0.0        0.0        0.0      10      0.0        0   
4  17.347895         0.0        0.0        0.0      17      0.0        0   

   cityA08U  cityCD  cityE  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
1       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
2       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
3       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
4       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   

                      createdOn                    modifiedOn  startStop  \
0  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
1  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
2  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
3  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
4  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   

   phevCity  phevHwy  phevComb  
0         0        0         0  
1         0        0         0  
2         0        0         0  
3         0        0         0  
4         0        0         0  

[5 rows x 83 columns]

In [7]:
data.shape

(37843, 83)

## Missing Values

From this initial view, we can see that our data set contains some columns that have missing values in them and others that seem to have a lot of zero values. Let's see how prevalent missing values are in our data. We can use the Pandas isnull method to check whether the value in each field is missing (null) and return either True or False for each field. We can use the sum method to total up the number of True values by column, and then we can add a condition using square brackets that will filter the data and show us only columns where the number of null values were greater than zero.

In [8]:
data.isnull()

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0          False       False      False      False   False    False    False   
1          False       False      False      False   False    False    False   
2          False       False      False      False   False    False    False   
3          False       False      False      False   False    False    False   
4          False       False      False      False   False    False    False   
...          ...         ...        ...        ...     ...      ...      ...   
37838      False       False      False      False   False    False    False   
37839      False       False      False      False   False    False    False   
37840      False       False      False      False   False    False    False   
37841      False       False      False      False   False    False    False   
37842      False       False      False      False   False    False    False   

       cityA08U  cityCD  cityE  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0         False   False  False  ...     True      True       False       True   
1         False   False  False  ...     True      True       False       True   
2         False   False  False  ...     True      True       False       True   
3         False   False  False  ...     True      True       False       True   
4         False   False  False  ...     True      True       False       True   
...         ...     ...    ...  ...      ...       ...         ...        ...   
37838     False   False  False  ...     True      True       False       True   
37839     False   False  False  ...     True      True       False       True   
37840     False   False  False  ...     True      True       False       True   
37841     False   False  False  ...     True      True       False       True   
37842     False   False  False  ...     True      True       False       True   

       createdOn  modifiedOn  startStop  phevCity  phevHwy  phevComb  
0          False       False       True     False    False     False  
1          False       False       True     False    False     False  
2          False       False       True     False    False     False  
3          False       False       True     False    False     False  
4          False       False       True     False    False     False  
...          ...         ...        ...       ...      ...       ...  
37838      False       False       True     False    False     False  
37839      False       False       True     False    False     False  
37840      False       False       True     False    False     False  
37841      False       False       True     False    False     False  
37842      False       False       True     False    False     False  

[37843 rows x 83 columns]

In [9]:
null_cols = data.isnull().sum()
null_cols[null_cols > 0] #If a sum value is 0 it means no value it missing. Here we see columns with a lot of empty values


cylinders       123
displ           120
drive          1189
eng_dscr      15403
trany            11
guzzler       35562
trans_dscr    22796
tCharger      32657
sCharger      37177
atvType       34771
fuelType2     36435
rangeA        36440
evMotor       37281
mfrCode       30818
c240Dscr      37806
c240bDscr     37807
startStop     31705
dtype: int64

We can see that some columns have relatively few null values while others have tens of thousands of nulls. For fields that have a lot of null values, you will often have to make a judgement call. If you don't think the information is going to be very useful to your analysis, then you would remove those columns from your data frame. In Pandas, we can do that using the drop method. For our purposes, let's remove the columns that have more than 10,000 null values in them. We will add these column names to a list, and then we will pass those columns to the drop method and indicate that we want columns (not rows) dropped by setting the axis parameter to 1.

In [12]:
null_cols[null_cols > 10000].index #.index makes a list from the object

Index(['eng_dscr', 'guzzler', 'trans_dscr', 'tCharger', 'sCharger', 'atvType',
       'fuelType2', 'rangeA', 'evMotor', 'mfrCode', 'c240Dscr', 'c240bDscr',
       'startStop'],
      dtype='object')

In [ ]:
drop_cols = list(null_cols[null_cols > 10000].index)
data = data.drop(drop_cols, axis=1) #1 is delete column wise

This leaves us with just a handful of remaining columns that have null values. Of the columns that remain, it looks like the cylinders column and the displ column have a similar number of nulls. Perhaps they are missing for similar reasons. We can investigate this by subsetting the data set and looking at just the records where displ is null and just the columns we think will be informative in allowing us to determine a reason.

In [16]:
data.shape #of columns reduced from 83 to 70 now

(37843, 70)

In [17]:
null_displ = data[(data['displ'].isnull()==True)] #See which cells are empty in selected rows. 
null_displ = null_displ[['year', 'make', 'model', 'trany', 'drive','fuelType','cylinders', 'displ']] #Apply this formula to all concerned columns
null_displ
#NaN in cylinders makes sense as electric cars don't have cylinders. So it's not corrupted.


year    make                            model           trany  \
7138   2000  Nissan                         Altra EV             NaN   
7139   2000  Toyota                          RAV4 EV             NaN   
8143   2001  Toyota                          RAV4 EV             NaN   
8144   2001    Ford                            Th!nk             NaN   
8146   2001    Ford           Explorer USPS Electric             NaN   
...     ...     ...                              ...             ...   
30969  2017     Kia                    Soul Electric  Automatic (A1)   
30972  2016   Tesla  Model S (60 kW-hr battery pack)  Automatic (A1)   
30973  2016   Tesla                Model S AWD - 60D  Automatic (A1)   
30974  2016   Tesla              Model S AWD - P100D  Automatic (A1)   
30975  2016   Tesla                Model X AWD - 60D  Automatic (A1)   

                   drive     fuelType  cylinders  displ  
7138                 NaN  Electricity        NaN    NaN  
7139       2-Wheel Drive  Electricity        NaN    NaN  
8143       2-Wheel Drive  Electricity        NaN    NaN  
8144                 NaN  Electricity        NaN    NaN  
8146       2-Wheel Drive  Electricity        NaN    NaN  
...                  ...          ...        ...    ...  
30969  Front-Wheel Drive  Electricity        NaN    NaN  
30972   Rear-Wheel Drive  Electricity        NaN    NaN  
30973    All-Wheel Drive  Electricity        NaN    NaN  
30974    All-Wheel Drive  Electricity        NaN    NaN  
30975    All-Wheel Drive  Electricity        NaN    NaN  

[120 rows x 8 columns]

We can see that most of the time, cylinders is null when displ is null and that the most of the records where both fields are null have a fuel type of Electricity. This makes sense, as electric cars do not have cylinders and can therefore not have any displacement. In this case, it would make sense to replace these null values with zeros. Pandas makes it easy to do that with the fillna method.

In [18]:
data[['displ', 'cylinders']] = data[['displ', 'cylinders']].fillna(0)
#edit the data to replace na values with 0. fillna = fill empty values

In [20]:
null_displ = data[(data['displ'].isnull()==True)] #Now no empty rows in 'displ' column
null_displ

Empty DataFrame
Columns: [barrels08, barrelsA08, charge120, charge240, city08, city08U, cityA08, cityA08U, cityCD, cityE, cityUF, co2, co2A, co2TailpipeAGpm, co2TailpipeGpm, comb08, comb08U, combA08, combA08U, combE, combinedCD, combinedUF, cylinders, displ, drive, engId, feScore, fuelCost08, fuelCostA08, fuelType, fuelType1, ghgScore, ghgScoreA, highway08, highway08U, highwayA08, highwayA08U, highwayCD, highwayE, highwayUF, hlv, hpv, id, lv2, lv4, make, model, mpgData, phevBlended, pv2, pv4, range, rangeCity, rangeCityA, rangeHwy, rangeHwyA, trany, UCity, UCityA, UHighway, UHighwayA, VClass, year, youSaveSpend, charge240b, createdOn, modifiedOn, phevCity, phevHwy, phevComb]
Index: []

[0 rows x 70 columns]

In this example, we filled the nulls in with zeros, but there are other strategies for filling in nulls. Depending on the circumstances, you might want to replace nulls with the column mean or mode values. Once you get more advanced, you can even use a variety of predictive imputation methods.

## Incorrect Values
In addition to null values, we also want to try to identify any values that seem incorrect. For example, in the previous section, we learned that a vehicle without cylinders should not have displacement and vice versa. Let's check to see if there are any cases that violate these rules.

In [23]:
test = data[(data['cylinders']==0) & (data['displ']!=0)]
test[['year', 'make', 'model', 'trany', 'drive','fuelType','cylinders', 'displ']]
#If you just put test it will return the data with all column headers. Here we have put test with column filters

year   make model         trany             drive fuelType  cylinders  \
21506  1986  Mazda  RX-7  Manual 5-spd  Rear-Wheel Drive  Regular        0.0   

       displ  
21506    1.3

Here we have identified a vehicle with a regular gasoline engine that reportedly does not have any cylinders but does have a value for displacement. The way we would correct this would be to either perform some domain research or ask a domain expert to find out how many actual cylinders this vehicle had. Alternatively, you can also try to look at similar vehicles in the data set and determine the most likely value for this field.

Suppose that using one of the aforementioned methods, we found out that this vehicle actually has a 4 cylinder engine. Once we have this information, we can use the loc method to update that specific value in the data frame.

In [ ]:
data.loc[(data['cylinders']==0) & (data['displ']!=0), 'cylinders'] = 4 #Update syntax with loc

## Low Variance Columns

When analyzing data, we want the fields we are working with to be informative, and we will want to strip away any columns that don't have a lot of value to us. One easy way to do this is to identify columns that have low variance, where the majority of the values in the column are the same. Since there is not a lot of variability in these columns, they have the potential to not be as informative as columns that have a variety of different values in them.

Let's try to identify columns where at least 90% of the values are the same so that we can remove them from our data set. To do this, we are going to create an empty list called low_variance that will eventually contain the names of columns that fit our criteria. We will then write a for loop that will take the minimum and the 90th percentile value for all the numeric columns in our data set (identified via the _get_numeric_data method). If the 90th percentile and the minimum are equal to each other, that means that at least 90% of the values in that column are the same and we will append that column name to our low_variance list.

In [ ]:
low_variance = []

for col in data._get_numeric_data(): #Limits to columns with numeric values
    minimum = min(data[col]) 
    ninety_perc = np.percentile(data[col], 90)
    if ninety_perc == minimum: 
        low_variance.append(col)

print(low_variance)

This returned 34 columns that we could potentially eliminate due to not having high enough variability to be informative. Of course, before we do this, we should check the values that do exist in these fields to confirm that they are not very informative. Once they have been checked, we can use the the drop method like we did earlier in this lesson to remove those columns from our data frame.

In [ ]:
data = data.drop(low_variance, axis=1)

## Extreme Values and Outliers

Now that we have removed low variance columns, we should look for outliers, or extreme values, in the columns that remain. These outliers can influence our aggregations when we are analyzing data later, so we want to make sure we address them during our data cleaning stage.

A common method for identifying outliers is one that leverages the interquartile range (IQR). Once the IQR is calculated, it is multiplied by a constant (typically 1.5) and lower and upper bounds are established at:

- 25th Percentile - (IQR x 1.5)
- 75th Percentile + (IQR x 1.5)

Any values outside this range are potential outliers and should be investigated.

Let's look at how we would do this for our data set using Python. We will use the Pandas describe function to easily calculate the 25th and 75th percentiles for every column and transpose the results so that we can easily reference the values in calculating the interquartile ranges.

In [28]:
stats = data.describe().transpose()

In [29]:
stats = data.describe().transpose()
stats['IQR'] = stats['75%'] - stats['25%']
stats

count          mean           std       min         25%  \
barrels08        37843.0     17.532506      4.575950      0.06    14.33087   
barrelsA08       37843.0      0.216169      1.141527      0.00     0.00000   
charge120        37843.0      0.000000      0.000000      0.00     0.00000   
charge240        37843.0      0.023531      0.427647      0.00     0.00000   
city08           37843.0     17.941389      6.660360      6.00    15.00000   
city08U          37843.0      4.042737      9.645820      0.00     0.00000   
cityA08          37843.0      0.520149      3.837874      0.00     0.00000   
cityA08U         37843.0      0.327163      3.542596      0.00     0.00000   
cityCD           37843.0      0.000406      0.039918      0.00     0.00000   
cityE            37843.0      0.184790      2.904558      0.00     0.00000   
cityUF           37843.0      0.000706      0.019458      0.00     0.00000   
co2              37843.0     61.503713    153.387715     -1.00    -1.00000   
co2A             37843.0      5.349919     55.539497     -1.00    -1.00000   
co2TailpipeAGpm  37843.0     17.771433     94.129283      0.00     0.00000   
co2TailpipeGpm   37843.0    473.179736    122.188847      0.00   388.00000   
comb08           37843.0     20.195809      6.623444      7.00    17.00000   
comb08U          37843.0      4.549751     10.389994      0.00     0.00000   
combA08          37843.0      0.581191      3.966255      0.00     0.00000   
combA08U         37843.0      0.359027      3.589260      0.00     0.00000   
combE            37843.0      0.190575      2.979567      0.00     0.00000   
combinedCD       37843.0      0.000335      0.035284      0.00     0.00000   
combinedUF       37843.0      0.000691      0.019049      0.00     0.00000   
cylinders        37843.0      5.719340      1.779180      0.00     4.00000   
displ            37843.0      3.307658      1.371982      0.00     2.20000   
engId            37843.0   8860.308961  17829.683477      0.00     0.00000   
feScore          37843.0     -0.035145      2.379584     -1.00    -1.00000   
fuelCost08       37843.0   1882.060090    510.280408    550.00  1500.00000   
fuelCostA08      37843.0     79.070105    417.668580      0.00     0.00000   
ghgScore         37843.0     -0.037206      2.374338     -1.00    -1.00000   
ghgScoreA        37843.0     -0.929630      0.627565     -1.00    -1.00000   
highway08        37843.0     24.104881      6.963192      9.00    20.00000   
highway08U       37843.0      5.438467     11.936327      0.00     0.00000   
highwayA08       37843.0      0.686891      4.308934      0.00     0.00000   
highwayA08U      37843.0      0.418698      3.808877      0.00     0.00000   
highwayCD        37843.0      0.000233      0.029564      0.00     0.00000   
highwayE         37843.0      0.198247      3.091862      0.00     0.00000   
highwayUF        37843.0      0.000673      0.018583      0.00     0.00000   
hlv              37843.0      2.026214      5.947989      0.00     0.00000   
hpv              37843.0     10.424332     28.148770      0.00     0.00000   
id               37843.0  19019.286235  11034.784855      1.00  9461.50000   
lv2              37843.0      1.849219      4.429983      0.00     0.00000   
lv4              37843.0      6.165658      9.743297      0.00     0.00000   
pv2              37843.0     13.696113     31.269930      0.00     0.00000   
pv4              37843.0     33.769125     45.914462      0.00     0.00000   
range            37843.0      0.392675      8.251191      0.00     0.00000   
rangeCity        37843.0      0.352618      8.049082      0.00     0.00000   
rangeCityA       37843.0      0.036800      1.158270      0.00     0.00000   
rangeHwy         37843.0      0.343399      8.171939      0.00     0.00000   
rangeHwyA        37843.0      0.033475      1.032642      0.00     0.00000   
UCity            37843.0     22.587229      9.350163      0.00    18.00000   
UCityA           37843.0      0.652380   

We will then create an empty data frame called outliers with the same columns as our data set. Finally, we will loop through each column in the data calculating the lower and upper bounds, retrieving records where the value for that column falls outside the bounds we established, and appending those results to our outlier data frame.

In [ ]:
outliers = pd.DataFrame(columns=data.columns)

for col in stats.index:
    iqr = stats.at[col,'IQR']
    cutoff = iqr * 1.5
    lower = stats.at[col,'25%'] - cutoff
    upper = stats.at[col,'75%'] + cutoff
    results = data[(data[col] < lower) | 
                   (data[col] > upper)].copy()
    results['Outlier'] = col
    outliers = outliers.append(results)


In [31]:
outliers

barrels08  barrelsA08  charge120  charge240 city08  city08U cityA08  \
47     32.961000    0.000000        0.0        0.0      9   0.0000       0   
58     32.961000    0.000000        0.0        0.0     10   0.0000       0   
69     32.961000    0.000000        0.0        0.0     10   0.0000       0   
80     32.961000    0.000000        0.0        0.0      9   0.0000       0   
275    32.961000    0.000000        0.0        0.0      9   0.0000       0   
...          ...         ...        ...        ...    ...      ...     ...   
30558   9.535201    0.426000        0.0        3.0     21  20.8000      47   
30833   8.009330    5.150156        0.0        3.0     25  24.8563      62   
30834   3.880578    3.329394        0.0        2.7     38  37.8163      99   
30873   7.071000    4.642394        0.0        2.0     28  27.7708      66   
30970   9.179989    5.885893        0.0        3.0     23  22.9620      55   

       cityA08U  cityCD  cityE  ...                              VClass  year  \
47       0.0000     0.0    0.0  ...                                Vans  1985   
58       0.0000     0.0    0.0  ...                                Vans  1985   
69       0.0000     0.0    0.0  ...                                Vans  1985   
80       0.0000     0.0    0.0  ...                                Vans  1985   
275      0.0000     0.0    0.0  ...              Standard Pickup Trucks  1993   
...         ...     ...    ...  ...                                 ...   ...   
30558   47.0475     0.0   71.0  ...  Standard Sport Utility Vehicle 4WD  2017   
30833   62.0000     0.0   53.0  ...                          Large Cars  2017   
30834   99.1000     0.0   34.0  ...                        Midsize Cars  2017   
30873   66.0249     0.0   51.0  ...                        Compact Cars  2017   
30970   55.0380     0.0   62.0  ...  Standard Sport Utility Vehicle 4WD  2017   

      youSaveSpend  charge240b                     createdOn  \
47          -10000         0.0  Tue Jan 01 00:00:00 EST 2013   
58          -10000         0.0  Tue Jan 01 00:00:00 EST 2013   
69          -10000         0.0  Tue Jan 01 00:00:00 EST 2013   
80          -10000         0.0  Tue Jan 01 00:00:00 EST 2013   
275         -10000         0.0  Tue Jan 01 00:00:00 EST 2013   
...            ...         ...                           ...   
30558        -1500         0.0  Wed Jul 20 00:00:00 EDT 2016   
30833            0         0.0  Wed Aug 17 00:00:00 EDT 2016   
30834         3000         0.0  Wed Aug 17 00:00:00 EDT 2016   
30873          750         0.0  Thu Aug 25 00:00:00 EDT 2016   
30970         -750         0.0  Tue Sep 13 00:00:00 EDT 2016   

                         modifiedOn  phevCity phevHwy  phevComb    Outlier  
47     Tue Jan 01 00:00:00 EST 2013         0       0         0  barrels08  
58     Tue Jan 01 00:00:00 EST 2013         0       0         0  barrels08  
69     Tue Jan 01 00:00:00 EST 2013         0       0         0  barrels08  
80     Tue Jan 01 00:00:00 EST 2013         0       0         0  barrels08  
275    Tue Jan 01 00:00:00 EST 2013         0       0         0  barrels08  
...                             ...       ...     ...       ...        ...  
30558  Wed Jul 20 00:00:00 EDT 2016        25      29        27   phevComb  
30833  Thu Aug 25 00:00:00 EDT 2016        31      37        33   phevComb  
30834  Wed Aug 17 00:00:00 EDT 2016        57      59        58   phevComb  
30873  Thu Aug 25 00:00:00 EDT 2016        34      43        38   phevComb  
30970  Tue Sep 13 00:00:00 EDT 2016        28      31        29   phevComb  

[94107 rows x 71 columns]

Our outliers data frame should now be populated with records that you can investigate further and determine whether they should be kept in the data or dropped. The Outlier column we added before appending the results for the column to the outliers data frame will let you know what column in each record contained the outlier. If you find that this method is returning too many results, you can be more stringent with your cutoff criteria (e.g. increasing the constant by which you multiply the IQR to 3 instead of 1.5).



## Data Type Correction


One common problem that is often overlooked is incorrect data types. This typically occurs when there is a numeric variable that should actually be represented as a categorical variable. The way to check the data type of each column in Pandas is by using the dtypes method.

In [32]:
#pd
data.dtypes

barrels08     float64
barrelsA08    float64
charge120     float64
charge240     float64
city08          int64
               ...   
createdOn      object
modifiedOn     object
phevCity        int64
phevHwy         int64
phevComb        int64
Length: 70, dtype: object

Pandas currently has the year column stored as integers, but what if we wanted the year to be stored as a categorical variable (object) instead? We could easily change that data type using the astype method and then check that it changed using the dtypes method again just on that field.

In [36]:
data['year'] = data['year'].astype('object')
data['year'].dtype

dtype('O')

You can apply this technique to any column whose data type you would like to change.



## Cleaning Text and Removing Special Characters


The presence of special characters in our fields has the potential to make analyzing our data challenging. Imagine not being able to perform calculations on a numeric field because it was currently represented as an object data type due to the fact that it had a dollar sign ($) in it. Similarly, imagine having a categorical field where you could not group records that belong in the same group together because in one field you are grouping by, terms that refer to the same thing are sometimes hyphenated. In cases like this, it is necessary to remove special characters so that we can properly analyze the data.

In our vehicles data set, the trany field has several special characters (parentheses, hyphens, etc.). We can take a look at the unique values in this column by using the set function.

In [37]:
print(set(data['trany']))

{nan, 'Auto(AV-S7)', 'Automatic (A1)', 'Auto(AM6)', 'Manual 4-spd Doubled', 'Auto(AM-S8)', 'Auto(AM8)', 'Automatic (S5)', 'Auto(L3)', 'Manual 4-spd', 'Auto (AV-S6)', 'Auto (AV)', 'Automatic 8-spd', 'Automatic (A6)', 'Manual 5-spd', 'Automatic (S4)', 'Manual 7-spd', 'Automatic 4-spd', 'Automatic (AM6)', 'Auto(AM7)', 'Automatic (AV)', 'Automatic 7-spd', 'Automatic 6spd', 'Automatic (S9)', 'Auto(AM-S6)', 'Automatic 3-spd', 'Auto(AM-S7)', 'Auto(AM-S9)', 'Auto(AV-S8)', 'Manual(M7)', 'Auto(A1)', 'Auto(L4)', 'Automatic (AV-S6)', 'Auto (AV-S8)', 'Manual 6-spd', 'Automatic 6-spd', 'Automatic (variable gear ratios)', 'Automatic (S8)', 'Manual 5 spd', 'Automatic (S7)', 'Automatic (AM5)', 'Auto(AM5)', 'Automatic 9-spd', 'Auto(AV-S6)', 'Manual 3-spd', 'Automatic (S6)', 'Automatic 5-spd'}


We can see that there are instances that refer to the same thing, but would not get grouped together due to special characters (e.g. Automatic 6-spd and Automatic 6spd). So let's remove all hyphens from this column with the help of the str.replace method and then print unique values again to ensure they were removed.

In [38]:
data['trany'] = data['trany'].str.replace('-', '')
print(set(data['trany']))

{nan, 'Automatic (A1)', 'Manual 4spd Doubled', 'Auto(AM6)', 'Auto(AM8)', 'Automatic (S5)', 'Auto(L3)', 'Automatic 8spd', 'Auto (AV)', 'Automatic 3spd', 'Automatic (A6)', 'Auto(AVS6)', 'Automatic 9spd', 'Manual 6spd', 'Manual 4spd', 'Automatic (S4)', 'Automatic (AM6)', 'Automatic 6spd', 'Auto(AM7)', 'Auto(AVS7)', 'Automatic (AV)', 'Automatic (S9)', 'Auto(AMS9)', 'Auto(AVS8)', 'Auto (AVS6)', 'Manual 3spd', 'Auto(A1)', 'Manual(M7)', 'Auto(L4)', 'Automatic (AVS6)', 'Auto(AMS6)', 'Manual 5spd', 'Automatic (variable gear ratios)', 'Auto(AMS8)', 'Automatic 4spd', 'Manual 7spd', 'Automatic (S8)', 'Manual 5 spd', 'Automatic 7spd', 'Automatic (S7)', 'Automatic (AM5)', 'Auto(AM5)', 'Automatic 5spd', 'Auto (AVS8)', 'Auto(AMS7)', 'Automatic (S6)'}


You will also notice that in some cases Automatic is abbreviated to Auto and in other cases it is spelled out. We can make that more consistent by using the same technique. While we are at it, let's also attempt to remove parentheses and make spacing more consistent.

In [39]:
data['trany'] = data['trany'].str.replace('Automatic', 'Auto')
data['trany'] = data['trany'].str.replace('Auto\(', 'Auto ')
data['trany'] = data['trany'].str.replace('Manual\(', 'Manual ')
data['trany'] = data['trany'].str.replace('\(', '')
data['trany'] = data['trany'].str.replace('\)', '')
print(set(data['trany']))

{nan, 'Auto AM5', 'Manual 4spd Doubled', 'Auto S4', 'Auto L4', 'Auto AMS8', 'Auto AMS7', 'Auto AM8', 'Manual 6spd', 'Manual 4spd', 'Auto S6', 'Auto 9spd', 'Auto AM7', 'Auto AVS8', 'Auto 6spd', 'Auto S9', 'Auto S7', 'Manual M7', 'Manual 3spd', 'Auto A1', 'Auto 4spd', 'Auto A6', 'Auto AM6', 'Auto 5spd', 'Manual 7spd', 'Manual 5 spd', 'Auto L3', 'Auto S8', 'Auto AVS6', 'Auto S5', 'Auto AV', 'Auto AMS9', 'Auto 8spd', 'Auto AMS6', 'Auto AVS7', 'Auto 7spd', 'Manual 5spd', 'Auto 3spd', 'Auto variable gear ratios'}


/var/folders/1k/45vk_vqn0z174f0vwyd90vqr0000gn/T/ipykernel_1360/189685402.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  data['trany'] = data['trany'].str.replace('Auto\(', 'Auto ')
/var/folders/1k/45vk_vqn0z174f0vwyd90vqr0000gn/T/ipykernel_1360/189685402.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  data['trany'] = data['trany'].str.replace('Manual\(', 'Manual ')
/var/folders/1k/45vk_vqn0z174f0vwyd90vqr0000gn/T/ipykernel_1360/189685402.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  data['trany'] = data['trany'].str.replace('\(', '')
/var/folders/1k/45vk_vqn0z174f0vwyd90vqr0000gn/T/ipykernel_1360/189685402.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  data['trany'] = data['trany'].str.replace('\)', '')


As you can see, we now have no special characters, consistent naming, and proper spacing. We started out with 47 unique values in this column, and using this technique, we were able to reduce the number of unique values to 39.


## Finding and Removing Duplicates

The final topic we are going to cover in this lesson is how to identify and remove duplicate rows (or rows that refer to the same entity) in our data. When trying to identify duplicates, we will use the columns (or attributes) of the data to help us determine what entities are similar enough to be considered the same entity. We want to start with all the columns we currently have available to us and work our way toward a lesser number of attributes in an intuitive fashion. In this process, the act of dropping duplicated records is easy, but identifying the correct attributes for comparison and which records to drop is sometimes quite challenging.

The first thing we will do is attempt to drop any duplicate records, considering all the columns we currently have in the data set. Pandas provides us with the ability to do that via the drop_duplicates method. We will use the len method to calculate the number of rows in the data set both before and after removing duplicates and then print the number of rows dropped.

In [40]:
before = len(data)
data = data.drop_duplicates()
after = len(data)
print('Number of duplicate records dropped: ', str(before - after))

Number of duplicate records dropped:  0


This tells us that there were no records that matched exactly across all columns. However, if we reduce the number of columns in our data that we are interested in, we can try again and have a higher likelihood of finding duplicate records. In the example below, we will select a subset of columns, remove all other columns, and then use the drop_duplicates method to drop any duplicate records based on the remaining columns.

In [41]:
select_columns = ['make', 'model', 'year', 'displ', 'cylinders', 
                  'trany', 'drive', 'VClass','fuelType','barrels08', 
                  'city08', 'highway08', 'comb08', 'co2TailpipeGpm', 'fuelCost08']

data = data[select_columns].drop_duplicates()
after = len(data)
print('Number of duplicate records dropped: ', str(before - after))
#This drops duplicates from certain columns. The ones we are interested in.


Number of duplicate records dropped:  885


With the reduced number of columns, we were able to identify and drop 885 duplicate records.



## Summary


In this lesson, we have learned many techniques for cleaning our data that will make it easier to analyze later. We started the lesson by learning how to examine the data for potential issues. We quickly moved on to finding missing values and incorrect values and correcting them. From there, we learned how to remove low variance columns and identify potential outliers. After that, we learned how to change a column's data type and clean our categorical variables in addition to learning how to identify and remove duplicate records. The techniques you have learned in this lesson are ones that you will find yourself revisiting and utilizing almost every time you prepare to analyze a data set, as most data sets you will encounter will require some level of cleaning.